File: benches/lang-comparison/c/linked_list.c
// Singly linked list with basic operations
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef struct Node {
    int64_t value;
    struct Node *next;
} Node;

Node *node_new(int64_t val) {
    Node *n = (Node *)malloc(sizeof(Node));
    n->value = val;
    n->next = NULL;
    return n;
}

void list_push(Node *head, int64_t val) {
    Node *n = node_new(val);
    n->next = head->next;
    head->next = n;
}

int64_t list_len(const Node *head) {
    int64_t count = 0;
    const Node *cur = head->next;
    while (cur != NULL) {
        count++;
        cur = cur->next;
    }
    return count;
}

int64_t list_sum(const Node *head) {
    int64_t total = 0;
    const Node *cur = head->next;
    while (cur != NULL) {
        total += cur->value;
        cur = cur->next;
    }
    return total;
}

int main() {
    Node head = {0, NULL};
    for (int64_t i = 1; i <= 10; i++) {
        list_push(&head, i);
    }
    printf("%lld\n", list_len(&head));
    printf("%lld\n", list_sum(&head));
    return 0;
}

File: benches/lang-comparison/rust/linked_list.rs
// Singly linked list with basic operations
struct Node {
    value: i64,
    next: Option<Box<Node>>,
}

impl Node {
    fn new(val: i64) -> Self {
        Node { value: val, next: None }
    }
}

fn list_push(head: &mut Node, val: i64) {
    let mut new_node = Box::new(Node::new(val));
    new_node.next = head.next.take();
    head.next = Some(new_node);
}

fn list_len(head: &Node) -> i64 {
    let mut count = 0i64;
    let mut cur = &head.next;
    while let Some(node) = cur {
        count += 1;
        cur = &node.next;
    }
    count
}

fn list_sum(head: &Node) -> i64 {
    let mut total = 0i64;
    let mut cur = &head.next;
    while let Some(node) = cur {
        total += node.value;
        cur = &node.next;
    }
    total
}

fn main() {
    let mut head = Node::new(0);
    for i in 1..=10 {
        list_push(&mut head, i);
    }
    println!("{}", list_len(&head));
    println!("{}", list_sum(&head));
}
